explainer.md

```diff
@@ -0,0 +1,273 @@
+      e.stream.write(clickMessage);
+    };
+  }
+}
+
+async function readStreamUntilFin(stream) {
+  const buffers = [];
+  let bufferedSize = 0;
+  let finished = false;
+  while (stream.readable) {
+    await stream.waitForReadable(1);
+    const buffer = new Uint8Array(stream.readableAmount);
+    const read = stream.readInto(readBuffer);
+    bufferedSize += read.amount;
+    finished = read.finished;
+  }
+  if (!finished) {
+    return null;
+  }
+
+  let joinedBuffer = new Uint8Array(bufferedSize);
+  let joinedSize = 0;
+  for (let buffer of buffers) {
+    joinedBuffer.set(buffer, joinedSize);
+    joinedSize += buffer.byteLength;
+  }
+  return joinedBuffer.array;
+}
+```
+
+## Example of requesting over HTTP and receiving media pushed out-of-order and unreliably over the same network connection
+
+```javascript
+const transport = PooledHttpTransport.getDatagramTransport();
+if (transport) {
+  await fetch('http://example.com/babyshark');
+  const datagrams = await transport.receiveDatagrams();
+  for (let data of datagrams) {
+    if (data) {
+      // Process the data
+    }
+  };
+}
+```
+
+## Detailed design discussion
+
+Any WebTransport can provide any of the following capabilities (mixins):
+
+: Unidirectional streams
+:: Indefintely long streams of bytes in one direction with back pressure applied
+to the sender when either the receiver can't read quickly enough or when
+constrained by network capacity/congestions.  Useful for sending messages that
+do not expect a response.  In-order, reliable messaging can be achieved by
+sending many messages in a single stream. Out-of-order messaging can be achieved
+by sending one message per stream. unreliable messaging can be achieved by
+disabling retransmissions for a particular stream.
+
+: Bidirectional streams
+:: Like unidirectional streams, but in two directions.  Useful for sending
+messages that expect a response.
+
+: Datagrams
+:: Small, out-of-order, unreliable messages.  Useful for sending small
+out-of-order, unreliable messages with less API complexity and less network
+overhead than streams.
+
+A QuicTransport is a WebTransport that maps directly to QUIC streams and
+datagrams, which makes it easy to connect to servers that speak QUIC with
+minimum overhead.  It supports all of these capabilities.
+
+A PooledHttpTransport is a WebTransport that provides different subset of these
+capabilities depending on the underlying HTTP protocol (HTTP3 providing the
+widest support).  It has the advantages that HTTP and non-HTTP traffic can share
+the same network port and congestion control context.
+
+## Alternative designs considered
+
+- [WebRTC Data Channel](https://developer.mozilla.org/en-US/docs/Web/API/RTCDataChannel)
+can be used, but require that the server endpoint implement several protocols
+uncommonly found on servers (ICE, DTLS, and SCTP) and that the client
+application use a complex API designed for a very different use case.
+
+- Layering WebSockets over HTTP/3
```

Comment:
Maybe instead of putting these as bullet points it would look nicer as headers (###).

Reply:
I didn't know that worked!  Sweet.